test/integration/test_raw_api.py

```diff
@@ -550,6 +551,71 @@ def raw_api_test_helper(raw_api, should_cleanup_old_buckets):
             is_file_lock_enabled=False,
         )
 
+    # b2_100_continue
+    print('b2_100_continue')
+    file_name = 'test-100-continue.txt'
+    file_contents = b'hello world'
+    file_sha1 = hex_sha1_of_stream(io.BytesIO(file_contents), len(file_contents))
+
+    def set_100_header(fn):
+        def wrapper(url, headers, *args, **kwargs):
+            headers.update({
+                'Expect': '100-continue',
+            })
+            return fn(url, headers, *args, **kwargs)
+
+        return wrapper
+
+    orig_send = http.client.HTTPConnection.send
+    sent_data = bytearray()
+
+    def patched_send(self, data):
+        sent_data.extend(data)
+        return orig_send(self, data)
+
+    with monkeypatch.context() as monkey:
+        monkey.setattr(
+            raw_api.b2_http,
+            'post_content_return_json',
+            set_100_header(raw_api.b2_http.post_content_return_json),
+        )
+        monkey.setattr(
+            http.client.HTTPConnection,
+            "send",
+            patched_send,
+        )
+        data = io.BytesIO(file_contents)
+
+        with pytest.raises(InvalidAuthToken):
+            raw_api.upload_file(
+                upload_url,
+                upload_auth_token + 'x',
+                file_name,
+                len(file_contents),
+                'text/plain',
+                file_sha1,
+                {'color': 'blue'},
+                data,
+                server_side_encryption=sse_b2_aes,
+            )
+
+        # Check if data was sent
+        assert file_contents not in sent_data
```

Comment:
I'm somewhat confused;
wouldn't it be probable that data wasn't sent just because invalidauthtoken error?

Reply:
I didn't get your question.

By default, `http.client` doesn't care if `Expect: 100-continue` header is set and sends the entire body with the request. When reading for response it just ignores `100 Continue` responses and keeps waiting for non-1XX response.

We are intercepting the request/response cycles when `Except: 100-continue` is present. So this test just verifies that we do not send the body initially, and we do not send the body at all if we do not receive `100 Continue` response.

Reply:
shouldn't it always fallback to sending data after a timeout?